### PR TITLE
Change task to allow a value for unmatched bonus

### PR DESCRIPTION
**Pull request by jonodrew**
### Changed
- The system now uses pickle under the hood, so please be careful - if you've not secured the connections between
  the machines you're running this on you could really get yourself into trouble.
- However, it has made it significantly faster - a single matching exercise is now down to just 40s, from a best of
  97s when we were using JSON to serialize data.

### Added

- This is the big one. We've added functionality that will creep up an `UnmatchedBonus`. This functionality is
  useful if you want to ensure everyone gets at least one mentor. It calculates a lot of values - one client is
  calculating 37 different iterations of a three-round program, requiring 111 rounds of matching - so it takes a bit
  longer to calculate. Exposing this functionality in the front end will be patched very soon, but in the meantime
  dig around in the routes section or add a `"pairing": True` key-value pair to your JSON call to the appropriate
  endpoint.
- Given the huge amount of processing happening, this functionality takes a lot longer than you're expecting. It's
  enough time to make several cups of tea - on my hardware, it's clocking in at around 7 or 8 minutes. That's a long
  time to stare at the same screen. We'll be updating the frontend to give more feedback soon, but for the moment,
  either check the logs from celery or accept that you'll be here a little while.
- A note about the approach: I could have built a system that iterated over potential outcomes sequentially,
  stopping when it got to the approach that scored above a specific threshold. I see two problems with this. First,
  assuming that each matching process takes _n_ seconds, in the worst case iterating upwards takes _Mn_ seconds. In
  the best case, of course, it takes _n_ seconds!
- My approach batches up the number of approaches into chunks of ten (_M_/10) that are done simultaneously (_Mn_/10).
  This is therefore generally faster, although not in the case where the first outcome is the one we want. Given
  that I can't predict things will be perfect every time, I've opted for the apparently longer approach.

### CODE REVIEW

**Comment by johnpeart** [comment]
Sorry for the delay.

I agree, adding a page in would be a good way of managing this. The page could be expanded, in future, to add other customisable variables.

I also like the idea of giving people human readable options to select, and letting the system working out the best weighting. It seems a more 'user centered' option, to me at least.

Do you need me to do anything specific to enable this?

**Comment by jonodrew** [comment]
If we take the second option, I'll need a wee form for the user to select their preferred option. That's all I need from you.

For me, I'll need to work out how to run multiple attempts separately. A very interesting engineering challenge!

**Comment by jonodrew** [comment]
It shouldn't need any js - as long as we use the design system we can keep using `<form>....</form>`

I'm going to try and work out the backend this week. We're aiming for the highest total score (measured by summing the successful `Match` score) that also results in every Mentee getting at least one Mentor

File: app/classes.py [hunk 1]
```diff
@@ -44,18 +44,24 @@ def val_grade_to_str(cls, grade: int):
...
         data["role"] = data["job title"]
-        data["email"] = data["email address"]
+        data["email"] = data.get("email address", data.get("email"))
         data["grade"] = CSPerson.str_grade_to_val(data.get("grade", ""))
+        return data
 
     def map_model_to_output(self, data: dict):
         data["job title"] = data.pop("role")
-        data["email address"] = data.pop("email")
+        # data["email address"] = data.pop("email")
```
Comment (jonodrew) [inline app/classes.py:60]: Note to self - fix this

**Comment by jonodrew** [comment]
@johnpeart 

Forgive the mess of commits - I'm going to redo them so it's clearer what's actually happening here.

The important part is [this line](https://github.com/jonodrew/mentor-match/blob/6f6fdee4b8b0676aa4d4d896a3159c66f1d19be7/app/main/routes.py#L119) - the system now expects a true or false value passed to it. It's where we use the javascript to start the matching process, if you remember?

True will kick off the the divining process, where it spins up every possible iteration of the entire, three-round grid. It's an absolutely mammoth task that takes almost ten minutes to run.

**Comment by jonodrew** [comment]
Screenshot 2022-06-18 at 21 40 11

Result of 6 minutes of work on the MacBook - the ideal, mentor-focussed outcome with the sample_data

**Comment by johnpeart** [comment]
Quite a lot of this is going over my head! 

But if I understand it at a rudimentary level, we want the form that toggles the 'true' or 'false' value to appear on the `process.html` page template, where I've added the commented out HTML in commit [910534c](https://github.com/jonodrew/mentor-match/pull/91/commits/910534c193f888ccae40715efba17e564e8125bc).

Is that right?

**Comment by jonodrew** [comment]
Perhaps. I think the issue we have is that when the user clicks the button,
some Javascript runs. If there's a form on the page the JS will need to
pick the answer off it, and I'm not sure how to do that. If you can work it
out I think that's the preferred method.

My approach would be a new page before this one with the standard form, and
then saving that value to the cookie/passing it back to the frontend
somehow.

**Comment by johnpeart** [comment]
That makes sense – though that is probably beyond my knowledge to do the routes, etc. I can make it look nice though!

Capturing the value of a radio button is relatively trivial in JS, so if you want me to do that instead, then I can do that.

**Comment by jonodrew** [comment]
I think we introduce a risk, because if a user has Javascript turned off it
might not work...?

It'll take a little longer but we should do it properly. If you can work up
a new HTML page in a fresh PR I'll fix up the routes to put it in the user
journey

Can't wait for you to try out this new functionality!

**Comment by johnpeart** [comment]
I've added some _very rough_ routes and a page template which I have put at `/options`.
It's in the [add-options-page](https://github.com/jonodrew/mentor-match/tree/add-options-page) branch.

Hopefully that helps?!

**Pull request merged**